src/Compilers/Core/Portable/SourceGeneration/Nodes/TransformNode.cs

```diff
@@ -14,39 +14,47 @@ namespace Microsoft.CodeAnalysis
     internal sealed class TransformNode<TInput, TOutput> : IIncrementalGeneratorNode<TOutput>
     {
         private readonly Func<TInput, ImmutableArray<TOutput>> _func;
+        private readonly IEqualityComparer<TOutput> _comparer;
         private readonly IIncrementalGeneratorNode<TInput> _sourceNode;
 
-        public TransformNode(IIncrementalGeneratorNode<TInput> sourceNode, Func<TInput, TOutput> userFunc)
-            : this(sourceNode, userFunc: i => ImmutableArray.Create(userFunc(i)))
+        public TransformNode(IIncrementalGeneratorNode<TInput> sourceNode, Func<TInput, TOutput> userFunc, IEqualityComparer<TOutput>? comparer = null)
+            : this(sourceNode, userFunc: i => ImmutableArray.Create(userFunc(i)), comparer)
         {
         }
 
-        public TransformNode(IIncrementalGeneratorNode<TInput> sourceNode, Func<TInput, ImmutableArray<TOutput>> userFunc)
+        public TransformNode(IIncrementalGeneratorNode<TInput> sourceNode, Func<TInput, ImmutableArray<TOutput>> userFunc, IEqualityComparer<TOutput>? comparer = null)
         {
             _sourceNode = sourceNode;
             _func = userFunc;
+            _comparer = comparer ?? EqualityComparer<TOutput>.Default;
         }
 
-        // PROTOTYPE(source-generators):
-        public IIncrementalGeneratorNode<TOutput> WithComparer(IEqualityComparer<TOutput> comparer) => this;
+        public IIncrementalGeneratorNode<TOutput> WithComparer(IEqualityComparer<TOutput> comparer) => new TransformNode<TInput, TOutput>(_sourceNode, _func, comparer);
 
         public NodeStateTable<TOutput> UpdateStateTable(DriverStateTable.Builder builder, NodeStateTable<TOutput> previousTable, CancellationToken cancellationToken)
         {
-            // PROTOTYPE(source-generators):caching, faulted etc.
+            // grab the source inputs
+            var sourceTable = builder.GetLatestStateTableForNode(_sourceNode);
+            if (sourceTable.IsCompacted)
+            {
+                return previousTable;
+            }
+            if (sourceTable.IsFaulted)
+            {
+                return NodeStateTable<TOutput>.FromFaultedTable(sourceTable);
+            }
 
             // Semantics of a transform:
             // Element-wise comparison of upstream table
             // - Cached or Removed: no transform, just use previous values
             // - Added: perform transform and add
             // - Modified: perform transform and do element wise comparison with previous results
 
-            // grab the source inputs
-            var sourceTable = builder.GetLatestStateTableForNode(_sourceNode);
             var newTable = new NodeStateTable<TOutput>.Builder();
 
             foreach (var entry in sourceTable)
             {
-                if (entry.state == EntryState.Cached || entry.state == EntryState.Removed)
+                if ((entry.state == EntryState.Cached || entry.state == EntryState.Removed) && !previousTable.IsEmpty)
```

Comment:
The state tables are generally 'tell, don't ask' and this breaks that. I'd like to refactor it out to essentially be a set of 'Try...' calls, but this is good enough for now.

Reply:
It occurs if a generator is added after some generation pass has run, as the downstream node may be new, but the upstream might have values. Interestingly, I think that can only happen to in the input nodes (because the others are all private to the generator), which has different semantics. That may mean that the invariants *do* hold. You will never see an empty previous table if you have a cached element.

---
In reply to: [626148483](https://github.com/dotnet/roslyn/pull/53147#discussion_r626148483) [](ancestors = 626148483)